include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+NonlinearSolverSelector<VectorType>::NonlinearSolverSelector(
+  const AdditionalData &additional_data)
+  : additional_data(additional_data)
+{
+  data_transfer(additional_data);
+}
+
+template <typename VectorType>
+NonlinearSolverSelector<VectorType>::NonlinearSolverSelector(
+  const AdditionalData &additional_data,
+  const MPI_Comm &      mpi_communicator)
+  : additional_data(additional_data)
+  , mpi_communicator(mpi_communicator)
+{
+  data_transfer(additional_data);
+}
+
+
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::select(
+  const typename AdditionalData::SolverType &type)
+{
+  additional_data.solver_type = type;
+}
+
+template <typename VectorType>
+NonlinearSolverSelector<VectorType>::AdditionalData::AdditionalData(
+  const SolverType &      solver_type,
+  const SolutionStrategy &strategy,
+  const unsigned int      maximum_non_linear_iterations,
+  const double            function_tolerance,
+  const double            relative_tolerance,
+  const double            step_tolerance,
+  const unsigned int      anderson_subspace_size)
+  : solver_type(solver_type)
+  , strategy(strategy)
+  , maximum_non_linear_iterations(maximum_non_linear_iterations)
+  , function_tolerance(function_tolerance)
+  , relative_tolerance(relative_tolerance)
+  , step_tolerance(step_tolerance)
+  , anderson_subspace_size(anderson_subspace_size)
+{}
+
+#ifdef DEAL_II_WITH_TRILINOS
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::set_data(
+  const typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData
+    &                                         additional_data,
+  const Teuchos::RCP<Teuchos::ParameterList> &parameters)
+{
+  additional_data_nox = additional_data;
+  parameters_nox      = parameters;
+}
+#endif
+
+#ifdef DEAL_II_WITH_SUNDIALS
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::set_data(
+  const typename SUNDIALS::KINSOL<VectorType>::AdditionalData &additional_data)
+{
+  additional_data_kinsol = additional_data;
+}
+#endif
+
+
+template <typename VectorType>
+void
+NonlinearSolverSelector<VectorType>::solve_with_petsc(
+  VectorType & /*initial_guess_and_solution*/)
+{
+  AssertThrow(false,
+              ExcMessage("PETSc SNES requires you to use PETSc vectors."));
+}
+
+
+
+#ifdef DEAL_II_WITH_PETSC
+template <>
+void
+NonlinearSolverSelector<PETScWrappers::MPI::Vector>::solve_with_petsc(
+  PETScWrappers::MPI::Vector &initial_guess_and_solution)
+{
+  PETScWrappers::NonlinearSolver<PETScWrappers::MPI::Vector> nonlinear_solver(
+    additional_data_petsc_snes, mpi_communicator);
+
+  nonlinear_solver.residual = residual;
+
+  nonlinear_solver.setup_jacobian = setup_jacobian;
+
+  nonlinear_solver.solve_with_jacobian =
+    [&](const PETScWrappers::MPI::Vector &src,
+        PETScWrappers::MPI::Vector &      dst) -> int {
+    const double tolerance = 1e-5;
+    return solve_with_jacobian(src, dst, tolerance);
+  };
```

Comment:
The return type is already no longer used:
```suggestion
  nonlinear_solver.solve_with_jacobian =
    [&](const PETScWrappers::MPI::Vector &src,
        PETScWrappers::MPI::Vector &      dst) {
    const double tolerance = 1e-5;
    solve_with_jacobian(src, dst, tolerance);
  };
```